packages/compat/src/transaction.ts

```diff
@@ -0,0 +1,158 @@
+    }
+    const isSigner = message.isAccountSigner(accountIndex);
+    const isWritable = message.isAccountWritable(accountIndex);
+
+    const role = isSigner
+        ? isWritable
+            ? AccountRole.WRITABLE_SIGNER
+            : AccountRole.READONLY_SIGNER
+        : isWritable
+        ? AccountRole.WRITABLE
+        : AccountRole.READONLY;
+
+    return {
+        address: accountPublicKey.toBase58() as Base58EncodedAddress,
+        role,
+    };
+}
+
+// While the type `MessageCompiledInstruction` says data is a `Uint8Array`,
+// it is actually a Buffer in tests. Presumably this is a bug in the old web3js compile
+// message code. We will just assume it can be either here
+type MessageCompiledInstructionButDataMightBeABuffer = Omit<MessageCompiledInstruction, 'data'> & {
+    data?: Uint8Array | Buffer;
+};
+
+function convertInstruction(
+    message: VersionedMessage,
+    accountKeys: MessageAccountKeys,
+    instruction: MessageCompiledInstructionButDataMightBeABuffer
+): IInstruction {
+    const programAddressPublicKey = accountKeys.get(instruction.programIdIndex);
+    if (!programAddressPublicKey) {
+        // TODO coded error
+        throw new Error(`Could not find program address at index ${instruction.programIdIndex}`);
+    }
+
+    const accounts = instruction.accountKeyIndexes.map(accountIndex =>
+        convertAccount(message, accountKeys, accountIndex)
+    );
+
+    let data: Uint8Array | undefined;
+    const instructionData: Uint8Array | Buffer | undefined = instruction.data;
+    if (instructionData && instructionData.length > 0) {
+        if (instructionData instanceof Uint8Array) {
+            data = instructionData;
+        }
+        if (instructionData instanceof Buffer) {
+            data = new Uint8Array(instructionData.length);
+            instructionData.forEach((val, i) => {
+                data![i] = val;
+            });
+        }
+    }
+
+    return {
+        programAddress: programAddressPublicKey.toBase58() as Base58EncodedAddress,
+        ...(accounts.length ? { accounts } : {}),
+        ...(data ? { data } : {}),
+    };
+}
+
+function convertSignatures(
+    transaction: VersionedTransaction,
+    staticAccountKeys: PublicKey[]
+): ITransactionWithSignatures['signatures'] {
+    const signatures: Record<Base58EncodedAddress, Ed25519Signature> = {};
+
+    for (let i = 0; i < transaction.signatures.length; i++) {
+        const address = staticAccountKeys[i].toBase58() as Base58EncodedAddress;
+
+        // old web3js includes a fake all 0 signature if it hasn't been signed
+        // we don't do that for the new tx model. So just skip if it's all 0s
+        const allZeros = transaction.signatures[i].every(byte => byte === 0);
+        if (allZeros) continue;
+
+        const signature = transaction.signatures[i] as Ed25519Signature;
+        signatures[address] = signature;
+    }
+
+    return signatures;
+}
```

Comment:
If this is just for type coercion, can you just do:

```typescript
function convertSignatures(
    transaction: VersionedTransaction,
    staticAccountKeys: PublicKey[]
): ITransactionWithSignatures['signatures'] {
    return transaction.signatures.filter(s => !s.every(byte => byte === 0)) as Ed25519Signature[];
}
```

Reply:
Or something similar. That might not totally work

Reply:
It's not just type coercion, it also re-shapes it to a dictionary, keyed by public key. The old `VersionedTransaction` just has an array of signatures, and they map implicitly to the first N accounts of the transaction. So (in functional terms) we basically zip the first N static accounts with the N signatures.

So you need to maintain the original index, to get to the right account key for each one, which means you can't filter the all 0s out first.

I think there is a purely functional way to do it using reduce, because you can get the index in the callback. That should work here.

Reply:
Refactored in 78764a44b33449822b743b887676882687547f20 - I think that's probably better?

Reply:
Slick!!